### PR TITLE
Fix/tao 10137/validate api params

### PR DESCRIPTION
**Pull request by augustas**
Remote environments are validated before creating async tasks. API updated to return informative messages about invalid params.
 
Related to : 
https://oat-sa.atlassian.net/browse/TAO-10404
https://oat-sa.atlassian.net/browse/TAO-10405
https://oat-sa.atlassian.net/browse/TAO-10406

First this PR changes how API request parameters are validated, both parameters, `delivery-uri` and `remote-environments`, must be included in the request body, with non-empty value(s). Secondly the `RemotePublishingService` is updated, in order to validate all provided environments before creating any tasks for publishing. In this case either tasks are created for all provided environments, or none of them. 

#### Steps to reproduce and test
 - QA has prepared 3 great tickets to describe the incorrect behavior and how to verify the fix in detail
 - In summary, submitting request without some of the parameters, or with an empty/invalid value, should return `400` response and informative message. Endpoint address is `taoPublishing/api/deliveries/publish`. List of different cases:
     - without or empty parameter `delivery-uri`
     - without or empty parameter `remote-environments`
     - provided `delivery-uri` parameter for non existing delivery
     - provided `remote-environments` parameter for non existing or disabled remote environment 
 
Test commands :
 
```
sudo -u www-data ./vendor/bin/phpunit taoPublishing/test/unit/
```

### CODE REVIEW

File: controller/api/Deliveries.php [hunk 1]
```diff
@@ -192,17 +192,8 @@ public function publish(ServerRequest $request)
...
-            }
-            if (!array_key_exists(self::REST_REMOTE_ENVIRONMENTS, $requestData)) {
-                throw new common_exception_RestApi(__('Missing required parameter: `%s`', self::REST_REMOTE_ENVIRONMENTS), 400);
-            }
-            $deliveryUri = $requestData[self::REST_DELIVERY_URI];
-            $remoteEnvironmentUris = $requestData[self::REST_REMOTE_ENVIRONMENTS];
-            if (!is_array($remoteEnvironmentUris)) {
-                $remoteEnvironmentUris = [$remoteEnvironmentUris];
-            }
+            $deliveryUri = $this->getDeliveryParameter($requestData);
```
Comment (zagovorichev) [inline controller/api/Deliveries.php:195]: getDeliveryIdParameter?

File: controller/api/Deliveries.php [hunk 2]
```diff
@@ -220,4 +211,33 @@ public function publish(ServerRequest $request)
             $this->returnFailure($e);
         }
     }
+
+    private function getDeliveryParameter(?array $requestData): string
```
Comment (zagovorichev) [inline controller/api/Deliveries.php:215]: either $requestData should be an array or you have to check that this is array

File: controller/api/Deliveries.php [hunk 3]
```diff
@@ -220,4 +211,33 @@ public function publish(ServerRequest $request)
...
+            throw new common_exception_RestApi(__('Missing required parameter: `%s`', self::REST_DELIVERY_URI), 400);
+        }
+        if (empty($requestData[self::REST_DELIVERY_URI])) {
+            throw new common_exception_RestApi(__('Parameter `%s` should not be empty.', self::REST_DELIVERY_URI), 400);
+        }
+
+        return $requestData[self::REST_DELIVERY_URI];
+    }
+
+    private function getRemoteEnvironmentsParameter(?array $requestData): array
```
Comment (zagovorichev) [inline controller/api/Deliveries.php:227]: $requestData is array or your code won't work, correct?

File: controller/api/Deliveries.php (hunk 1 above)
Comment (augustas) [reply controller/api/Deliveries.php]: Renamed to `getDeliveryUriParameter`

File: controller/api/Deliveries.php (hunk 2 above)
Comment (augustas) [reply controller/api/Deliveries.php]: Correct, fixed

File: controller/api/Deliveries.php (hunk 3 above)
Comment (augustas) [reply controller/api/Deliveries.php:227]: Correct, fixed

File: controller/api/Deliveries.php (hunk 1 above)
Comment (zagovorichev) [reply controller/api/Deliveries.php]: about this URI: I'm OK with that, because all project uses it, but it is more like an ID, example: when we move it to KV storage it becomes kv_de_uri, so for delivery it could be kv_d_uri

**Comment by zagovorichev** [comment]
@augustas , could you provide todo list, to check all aspects of this PR? it would be really useful for history, and for reviewers of course too

**Comment by augustas** [comment]
> @augustas , could you provide todo list, to check all aspects of this PR? it would be really useful for history, and for reviewers of course too

Thank you, I have updated the description

File: controller/api/Deliveries.php [hunk 4]
```diff
@@ -192,17 +192,8 @@ public function publish(ServerRequest $request)
                 throw new common_exception_NotImplemented('Only POST method is accepted to publish deliveries');
             }
             $requestData = $request->getParsedBody();
```
[the remark this answers is not in the document]
Comment (augustas) [reply controller/api/Deliveries.php:193]: Done
